### PR TITLE
feat!: have oldVal passed on notify

### PR DESCRIPTION
**Pull request by y-nk**
:warning: this is a breaking change!

i'm opening this PR as a way to demonstrate a feature i'd like to have. it's not very heavy as you can see, and can help in many situations where we would like to compute diff between old and new value in userland.

currently, this is not possible easily. you need to hold the value at the end of a `.listen` and do the job manually, but i think this kind of api is very useful and comfortable (as in Vue the watchers have `(newVal?: T, oldVal?: T) => void`)

Current workaround:

```js
let oldVal = atom.get()

atom.subscribe(newVal => {
  const diff = newVal - oldVal
  oldVal = newVal
})
```

### CODE REVIEW

**Comment by ai** [comment]
I like the idea.

I rebased PR to `next` (the branch for next release). Let’s fix CI first:

1. Remove `,` at the end of the lists (could you install Prettier to text editor?)
2. Avoid code style changes
3. Update `size-limit` in `package.json` to see the price of the feature
4. Add tests

**Comment by y-nk** [comment]
I've been fixing lint + updated size-limit. The addition is:

+4B on atom
+9B on the whole package

Due to the size of the addition in atom, i cannot see a way to keep it under 300 😓 i've tried variations of `notify()` with `notify(...args)` or even `notify()` and leverage `arguments` but all came fatter than the original implementation.

If you agree with this, i'll add tests further on to make sure it's solid. Please let me know whenever you have time.

File: atom/index.js [hunk 1]
```diff
@@ -24,28 +24,29 @@ export let atom = (initialValue, level) => {
...
-               listenerQueue[i],
-               listenerQueue[i + 1],
-               listenerQueue[i + 2],
-               listenerQueue[i + 3]
-             )
+                listenerQueue[i],
+                listenerQueue[i + 1],
+                listenerQueue[i + 2],
+                listenerQueue[i + 3]
+              )
```
Comment (y-nk) [inline atom/index.js:49]: the spaces here are not my design, they were added when running `eslint --fix` 🤷

File: atom/index.js [hunk 2]
```diff
@@ -56,12 +55,15 @@ export let atom = (initialValue, level) => {
...
       }
     },
-    off() {}, /* It will be called on last listener unsubscribing.
-                 We will redefine it in onMount and onStop. */
-    set(data) {
-      if ($atom.value !== data) {
-        $atom.value = data
-        $atom.notify()
+
+    /* It will be called on last listener unsubscribing.We will redefine it in onMount and onStop. */
```
Comment (ai) [inline atom/index.js:59]: Could you revert comment changes?

File: deep-map/index.js [hunk 3]
```diff
@@ -5,10 +5,11 @@ export { getPath, setPath } from './path.js'
 
 export function deepMap(initial = {}) {
   let $deepMap = atom(initial)
-  $deepMap.setKey = (key, value) => {
-    if (getPath($deepMap.value, key) !== value) {
-      $deepMap.value = setPath($deepMap.value, key, value)
-      $deepMap.notify(key)
+  $deepMap.setKey = (key, newVal) => {
+    let oldVal = getPath($deepMap.value, key)
+    if (oldVal !== newVal) {
```
Comment (ai) [inline deep-map/index.js:10]: Could you use `newValue` and `oldValue` name? `val` looks like to cryptic for me.

File: deep-map/index.js [hunk 4]
```diff
@@ -5,10 +5,11 @@ export { getPath, setPath } from './path.js'
...
+      $deepMap.value = setPath($deepMap.value, key, newVal)
+      $deepMap.notify(oldVal, key)
```
Comment (y-nk) [inline deep-map/index.js:12]:
```suggestion
  $deepMap.setKey = (key, newValue) => {
    let oldValue = getPath($deepMap.value, key)
    if (oldValue !== newValue) {
      $deepMap.value = setPath($deepMap.value, key, newValue)
      $deepMap.notify(oldValue, key)
```

**Comment by y-nk** [comment]
@ai i have a few days ahead available ; do you still want it?

**Comment by gismya** [comment]
@y-nk: if you have issues finding the time I could adopt this PR and get it ready for merging? It's a feature that I'd love to have.

**Comment by gismya** [comment]
Would you prefer me to make a new fork or make me a collaborator on your?

**Comment by y-nk** [comment]
@gismya i've added you as collaborator. you can also fork it of course, so you're free to pick the solution you prefer :)

**Comment by gismya** [comment]
It was a bit further from being ready to merge than I thought. I'll pick up the baton, but doing it from a clean slate.

**Comment by ai** [comment]
Closing in favor of https://github.com/nanostores/nanostores/pull/269

**Pull request closed without merging**